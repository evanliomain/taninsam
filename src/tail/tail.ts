/**
 * @module array=>array
 */
/**
 * Extract the tail of an array and return an empty array if the original one is empty
 * @return the function to apply on the TBD to do something
 * @example
 * ```
 * tail<number>()([1, 2, 3, 4, 5]) // [2, 3, 4, 5]
 * ```
 * @example Using the chain
 * ```
 * chain([1, 2, 3, 4, 5])
 *   .chain(tail<number>())
 *   .value() // [2, 3, 4, 5]
 * ```
 */
export function tail<T>(): (array: ReadonlyArray<T>) => ReadonlyArray<T> {
  return (array: ReadonlyArray<T>) => {
    if (0 === array.length) {
      return [];
    }
    const [_, ...end] = array;
    return end;
  };
}
