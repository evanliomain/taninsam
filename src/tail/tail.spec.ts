import { describe, expect, test } from 'vitest';
import { tail } from './tail';

describe('tail function', () => {
  const input = [1, 2, 3, 4, 5];

  test('is a pure function with array', () => {
    tail()(input);
    expect(input).toMatchSnapshot();
  });

  test('[1, 2, 3, 4, 5] |> tail === [2, 3, 4, 5]', () => {
    expect(tail()(input)).toMatchSnapshot();
  });

  test('[] |> tail === []', () => {
    expect(tail()([])).toMatchSnapshot();
  });
});
