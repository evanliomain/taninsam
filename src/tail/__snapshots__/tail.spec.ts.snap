// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`tail function [] |> tail === [] 1`] = `Array []`;

exports[`tail function [1, 2, 3, 4, 5] |> tail === [2, 3, 4, 5] 1`] = `
Array [
  2,
  3,
  4,
  5,
]
`;

exports[`tail function is a pure function with array 1`] = `
Array [
  1,
  2,
  3,
  4,
  5,
]
`;
